social_core/backends/linkedin.py

```diff
@@ -2,94 +2,108 @@
+    USER_DETAILS_URL = \
+        'https://api.linkedin.com/v2/me?projection=({projection})'
+    USER_EMAILS_URL = 'https://api.linkedin.com/v2/emailAddress' \
+                      '?q=members&projection=(elements*(handle~))'
+    ACCESS_TOKEN_METHOD = 'POST'
+    REDIRECT_STATE = False
+    DEFAULT_SCOPE = ['r_liteprofile']
+    EXTRA_DATA = [
+        ('id', 'id'),
+        ('expires_in', 'expires'),
+        ('firstName', 'first_name'),
+        ('lastName', 'last_name')
+    ]
 
     def user_details_url(self):
         # use set() since LinkedIn fails when values are duplicated
-        fields_selectors = list(set(['first-name', 'id', 'last-name'] +
-                                self.setting('FIELD_SELECTORS', [])))
+        fields_selectors = list(set(['id', 'firstName', 'lastName'] +
+                                    self.setting('FIELD_SELECTORS', [])))
         # user sort to ease the tests URL mocking
         fields_selectors.sort()
         fields_selectors = ','.join(fields_selectors)
-        return self.USER_DETAILS.format(fields_selectors)
-
-    def user_data_headers(self):
-        lang = self.setting('FORCE_PROFILE_LANGUAGE')
-        if lang:
-            return {
-                'Accept-Language': lang if lang is not True
-                                        else self.strategy.get_language()
-            }
+        return self.USER_DETAILS_URL.format(projection=fields_selectors)
 
-
-class LinkedinOAuth(BaseLinkedinAuth, BaseOAuth1):
-    """Linkedin OAuth authentication backend"""
-    name = 'linkedin'
-    SCOPE_SEPARATOR = '+'
-    AUTHORIZATION_URL = 'https://www.linkedin.com/uas/oauth/authenticate'
-    REQUEST_TOKEN_URL = 'https://api.linkedin.com/uas/oauth/requestToken'
-    ACCESS_TOKEN_URL = 'https://api.linkedin.com/uas/oauth/accessToken'
+    def user_emails_url(self):
+        return self.USER_EMAILS_URL
 
     def user_data(self, access_token, *args, **kwargs):
-        """Return user data provided"""
-        return self.get_json(
+        response = self.get_json(
             self.user_details_url(),
-            params={'format': 'json'},
-            auth=self.oauth_auth(access_token),
-            headers=self.user_data_headers()
+            headers=self.user_data_headers(access_token)
         )
 
-    def unauthorized_token(self):
-        """Makes first request to oauth. Returns an unauthorized Token."""
-        scope = self.get_scope() or ''
-        if scope:
-            scope = '?scope=' + self.SCOPE_SEPARATOR.join(scope)
-        return self.request(self.REQUEST_TOKEN_URL + scope,
-                            params=self.request_token_extra_arguments(),
-                            auth=self.oauth_auth()).text
-
+        if 'emailAddress' in set(self.setting('FIELD_SELECTORS', [])):
+            emails = self.email_data(access_token, *args, **kwargs)
+            if emails:
+                response['emailAddress'] = emails[0]
```

Comment:
I hit an exception here. `TypeError: 'filter' object is not subscriptable`

Should it be `response['emailAddress'] = next(emails)]` ?

Reply:
I changed the 'return Filter(None, emailAddresses)' line in the 'email_data' function to just 'return emailAdresses' to work around this. @awachat, what is the use of the filter without a filterfunction?

Reply:
I think it's just to filter out `None`s.

Maybe call `list()` on the result before returning?

Reply:
It works in python2, looks like in python3 .filter() does not return a list but an iterator. 

`response['emailAddress'] = next(emails)` -> works only in python3 not in python2. 

Thanks for flagging this.

@W3Technics, `None` filter on a list removes all the falsify values from the list.
```
filter(None, [1, None, "", {}, []])

# return
[1]
```

If there is anything else breaking with python3, please do let me know.